cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,699 @@
+
+  pOp->pMqttConnection = pConn;
+
+  if ( pOp->incomingPublish ) {
+    // allocate publish member of the union
+
+    allocate_IotMqttPublishInfo( &(pOp->u.publish.publishInfo ) );
+
+    size_t length;
+    __CPROVER_assume( VALID_CBMC_SIZE( length ) );
+    pOp->u.publish.pReceivedData = malloc_can_fail( length );
+
+  } else {
+    // allocate operation member of the union
+
+    // assumption here is checked below in valid_*
+    __CPROVER_assume( VALID_CBMC_SIZE( pOp->u.operation.packetSize ) );
+    pOp->u.operation.pMqttPacket =
+      malloc_can_fail( pOp->u.operation.packetSize );
+
+  }
+
+  return pOp;
+}
+
+// TODO: valid_ should have same signature as allocate_,
+// should check that operation points to the connection
+
+bool valid_IotMqttOperation( const IotMqttOperation_t pOp )
+{
+  if ( pOp == NULL ) return false;
+
+  // publish member of the union should be valid (pOp->incomingPublish )
+
+  bool valid_publishinfo =
+    valid_IotMqttPublishInfo( &(pOp->u.publish.publishInfo) );
+
+  bool valid_publish_member =
+    IMPLIES( pOp->incomingPublish, valid_publishinfo );
+
+  // operation member of the union should be valid (!pOp->incomingPublish)
+
+  bool valid_packet =
+    VALID_STRING( pOp->u.operation.pMqttPacket, pOp->u.operation.packetSize ) &&
+#ifdef PACKET_SIZE_MAX
+    // Some proofs iterate over a packet and must bound the packet size
+    pOp->u.operation.packetSize < PACKET_SIZE_MAX &&
+#endif
+    VALID_CBMC_SIZE( pOp->u.operation.packetSize );
+  bool valid_pingreq_packet =
+    IMPLIES( pOp->u.operation.type == IOT_MQTT_PINGREQ,
+	     IFF( pOp->u.operation.periodic.ping.keepAliveMs == 0,
+		  pOp->u.operation.packetSize == 0 ) );
+  bool valid_other_packet =
+    IMPLIES( pOp->u.operation.type != IOT_MQTT_PINGREQ,
+	     pOp->u.operation.packetSize > 0 );
+  bool waitable =
+    ( pOp->u.operation.flags & IOT_MQTT_FLAG_WAITABLE )
+        == IOT_MQTT_FLAG_WAITABLE;
+  bool valid_jobReference =
+    // Async operations are waitable.  Loosely speaking, an async operation
+    // is split into independent send and ack events, and an sync operation
+    // is not.
+    IMPLIES(  waitable, pOp->u.operation.jobReference == 2 ) &&
+    IMPLIES( !waitable, pOp->u.operation.jobReference == 1 );
+
+  bool valid_operation_member =
+    IMPLIES( !pOp->incomingPublish,
+	     valid_packet &&
+	     valid_pingreq_packet &&
+	     valid_other_packet &&
+	     valid_jobReference );
+
+  return
+    // assume valid connection
+    valid_publish_member &&
+    valid_operation_member;
+}
+
+/****************************************************************
+ * IotMqttOperation list
+ ****************************************************************/
+
+IotListDouble_t *allocate_IotMqttOperationList( IotListDouble_t *pOp,
+						size_t length,
+						IotMqttConnection_t pConn )
+{
+  if ( pOp == NULL ) pOp = malloc_can_fail( sizeof( *pOp ) );
+  if ( pOp == NULL ) return NULL;
+
+  // Allocate lists of length L <= 3 (MAX = L+1)
```

Comment:
Comment explaining why this is sufficient for coverage?

Reply:
It is not.  It is a limitation of the proof.  We can, of course, make L as large as CBMC can tolerate, but there will always be an L and it will never be huge.